File: server/healthaggregation/healthaggregation/pseudorandomgenerator.py
from nacl.bindings import crypto_hash_sha256
from nacl.utils import randombytes_deterministic

java_long_byte_size = 8


class PseudoRandomGenerator:

    def __init__(self, seed: bytes):
        self.seed = int.from_bytes(seed, byteorder='big', signed=True)
        self.byte_length = len(seed)

    def __get_and_increment_seed__(self):
        seed_bytes = self.seed.to_bytes(self.byte_length, byteorder='big', signed=True)
        self.seed += 1
        # The LibSodium Pseudorandomgenerator only works with a 32 byte seed
        return crypto_hash_sha256(seed_bytes)[:32]

    def generate_random_bytes(self, size: int):
        return randombytes_deterministic(size, self.__get_and_increment_seed__())

    def generate_random_perturbation(self, length: int):
        perturbation = []
        for i in range(length):
            perturbation.append(int.from_bytes(self.generate_random_bytes(java_long_byte_size), byteorder='big', signed=True))
        return perturbation

File: client-android/app/src/main/java/com/strobel/healthaggregation/crypto/PseudorandomGenerator.java
package com.strobel.healthaggregation.crypto;

import com.goterl.lazysodium.LazySodiumAndroid;
import com.goterl.lazysodium.SodiumAndroid;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

public class PseudorandomGenerator {

    private static final String TAG = "PseudorandomGenerator";
    private static final LazySodiumAndroid sodium = new LazySodiumAndroid(new SodiumAndroid());
    private static final ByteBuffer byteBuffer = ByteBuffer.allocate(Long.BYTES);

    private BigInteger seed;

    public PseudorandomGenerator(byte [] seed) {
        this.seed = new BigInteger(seed);
    }

    private byte[] getAndIncrementSeed() {
        byte [] seed = this.seed.toByteArray();
        this.seed = this.seed.add(BigInteger.ONE);
        try {
            // The LibSodium Pseudorandomgenerator only works with a 32 byte seed
            return Arrays.copyOf(MessageDigest.getInstance("SHA-256").digest(seed), 32);
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
            return null;
        }
    }

    public byte[] generateRandomBytes(int x) {
        return sodium.randomBytesDeterministic(x, getAndIncrementSeed());
    }

    public long[] generateRandomPerturbation(int length) {
        long[] ret = new long[length];
        for(int i = 0; i < ret.length; i++) {
            byteBuffer.clear();
            byteBuffer.put(generateRandomBytes(Long.BYTES));
            byteBuffer.flip();
            ret[i] = byteBuffer.getLong();
        }
        return ret;
    }
}